providers/fips/fipsindicator.c

```diff
@@ -0,0 +1,113 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/indicator.h>
+#include <openssl/params.h>
+#include <openssl/core_names.h>
+#include "prov/fipsindicator.h"
+#include "internal/common.h" /* for ossl_assert() */
+
+void ossl_FIPS_IND_init(OSSL_FIPS_IND *ind)
+{
+    ossl_FIPS_IND_set_approved(ind); /* Assume we are approved by default */
+    memset(ind->settable, OSSL_FIPS_IND_STATE_UNKNOWN, sizeof(ind->settable));
+}
+
+void ossl_FIPS_IND_set_approved(OSSL_FIPS_IND *ind)
+{
+    ind->approved = 1;
+}
+
+void ossl_FIPS_IND_copy(OSSL_FIPS_IND *dst, const OSSL_FIPS_IND *src)
+{
+    *dst = *src;
+}
+
+void ossl_FIPS_IND_set_settable(OSSL_FIPS_IND *ind, int id, int state)
+{
+    if (!ossl_assert(id < OSSL_FIPS_IND_SETTABLE_MAX))
+        return;
+    if (!ossl_assert(state == OSSL_FIPS_IND_STATE_STRICT
+                     || state == OSSL_FIPS_IND_STATE_TOLERANT))
+        return;
+    ind->settable[id] = state;
+}
+
+int ossl_FIPS_IND_get_settable(const OSSL_FIPS_IND *ind, int id)
+{
+    if (!ossl_assert(id < OSSL_FIPS_IND_SETTABLE_MAX))
+        return OSSL_FIPS_IND_STATE_UNKNOWN;
+    return ind->settable[id];
+}
+
+/*
+ * This should only be called when a strict FIPS algorithm check fails.
+ * It assumes that we are in strict mode by default.
+ * If the logic here is not sufficient for all cases, then additional
+ * ossl_FIPS_IND_on_unapproved() functions may be required.
+ */
+int ossl_FIPS_IND_on_unapproved(OSSL_FIPS_IND *ind, int id,
+                                OSSL_LIB_CTX *libctx,
+                                const char *algname, const char *opname,
+                                OSSL_FIPS_IND_CHECK_CB *config_check_fn)
+{
+    /* Set to unapproved. Once unapproved mode is set this will not be reset */
+    ind->approved = 0;
+
+    /*
+     * We only trigger the indicator callback if the ctx variable is cleared OR
+     * the configurable item is cleared. If the values are unknown they are
+     * assumed to be strict.
+     */
+    if (ossl_FIPS_IND_get_settable(ind, id) == OSSL_FIPS_IND_STATE_TOLERANT
+        || (config_check_fn != NULL
+            && config_check_fn(libctx) == OSSL_FIPS_IND_STATE_TOLERANT)) {
+        return ossl_FIPS_IND_callback(libctx, algname, opname);
+    }
+    /* Strict mode gets here: This returns an error */
+    return 0;
+}
+
+int ossl_FIPS_IND_set_ctx_param(OSSL_FIPS_IND *ind, int id,
+                                const OSSL_PARAM params[], const char *name)
+{
+    int in = 0;
+    const OSSL_PARAM *p = OSSL_PARAM_locate_const(params, name);
+
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_int(p, &in))
+            return 0;
+        ossl_FIPS_IND_set_settable(ind, id, in);
+    }
+    return 1;
+}
+
+int ossl_FIPS_IND_get_ctx_param(const OSSL_FIPS_IND *ind, OSSL_PARAM params[])
+{
+    OSSL_PARAM *p = OSSL_PARAM_locate(params, OSSL_ALG_PARAM_FIPS_APPROVED_INDICATOR);
+
+    return p == NULL || OSSL_PARAM_set_int(p, ind->approved);
```

Comment:
If we use `OSSL_FIPS_IND_GET_CTX_PARAM` in `*kdf_get_ctx_params`,
this may make test `do_kdf_hkdf_gettables` broken?

Take HKDF as an example:
```c
static int kdf_hkdf_get_ctx_params(void *vctx, OSSL_PARAM params[])
{
    KDF_HKDF *ctx = (KDF_HKDF *)vctx;
    /* ... */
    if (!OSSL_FIPS_IND_GET_CTX_PARAM(ctx, params))
        return 0;
    return 1;
}
```
The above make [the assertion](https://github.com/openssl/openssl/blob/8f250985ad1ac4efc25621ce2504c52ef0cbe283/test/evp_kdf_test.c#L278) failed.

Reply:
This seems like a side issue that is not caused by this PR.

kdf_hkdf_get_ctx_params() already seems broken...
The way this is written is that only 1 param will be accepted at once.. (This seems to defeat the purpose of having a param[]). The pattern used elsewhere is to return 0 on failure OR return 1 otherwise, but you dont normally stop on the first param found..

````
static int kdf_hkdf_get_ctx_params(void *vctx, OSSL_PARAM params[])
{
    KDF_HKDF *ctx = (KDF_HKDF *)vctx;
    OSSL_PARAM *p;

    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_SIZE)) != NULL) {
        size_t sz = kdf_hkdf_size(ctx);

        if (sz == 0)
            return 0;
        return OSSL_PARAM_set_size_t(p, sz);
    }
    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_INFO)) != NULL) {
        if (ctx->info == NULL || ctx->info_len == 0) {
            p->return_size = 0;
            return 1;
        }
        return OSSL_PARAM_set_octet_string(p, ctx->info, ctx->info_len);
    }
    return -2;
}
````

Reply:
Okay, so the correct implementation  for current `kdf_hkdf_get_ctx_params()` should be like the following, is it right?

```c
static int kdf_hkdf_get_ctx_params(void *vctx, OSSL_PARAM params[])
{
    int ret = -2;
    KDF_HKDF *ctx = (KDF_HKDF *)vctx;
    OSSL_PARAM *p;

    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_SIZE)) != NULL) {
        size_t sz = kdf_hkdf_size(ctx);
        ret = 1;

        if (sz == 0)
            return 0;
        if (!OSSL_PARAM_set_size_t(p, sz))
            return 0;
    }
    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_INFO)) != NULL) {
        ret = 1;

        if (ctx->info == NULL || ctx->info_len == 0)
            p->return_size = 0;
        else if (!OSSL_PARAM_set_octet_string(p, ctx->info, ctx->info_len))
            return 0;
    }
    return ret;
}
```

Reply:
However, I though it is still unable to distinguish between `the request parameter is unsupported` and `OSSL_PARAM_set_int() operation failed` if we use `OSSL_FIPS_IND_GET_CTX_PARAM()`.

Therefore, the return value for `kdf_hkdf_get_ctx_params()` may be incorrect. Example is as the following.

```c
static int kdf_hkdf_get_ctx_params(void *vctx, OSSL_PARAM params[])
{
    int ret = -2;
    KDF_HKDF *ctx = (KDF_HKDF *)vctx;
    OSSL_PARAM *p;

    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_SIZE)) != NULL) {
        size_t sz = kdf_hkdf_size(ctx);
        ret = 1;

        if (sz == 0)
            return 0;
        if (!OSSL_PARAM_set_size_t(p, sz))
            return 0;
    }
    if ((p = OSSL_PARAM_locate(params, OSSL_KDF_PARAM_INFO)) != NULL) {
        ret = 1;

        if (ctx->info == NULL || ctx->info_len == 0)
            p->return_size = 0;
        else if (!OSSL_PARAM_set_octet_string(p, ctx->info, ctx->info_len))
            return 0;
    }
    /**
     * If all requested parameters in `params` are unsupported, the return value (1) is incorrect.
     * 
     * If we don't set `ret` to one, considering there is only
     * `OSSL_ALG_PARAM_FIPS_APPROVED_INDICATOR` in `params` and `OSSL_PARAM_set_int()` failed,
     * the return value (-2) is still incorrect.
     * 
     * ```
     * ...
     * 
     * if (!OSSL_FIPS_IND_GET_CTX_PARAM(((KDF_SSHKDF *)vctx), params))
     *     return 0;
     * return ret;
     * ```
     * 
     */
    if (OSSL_FIPS_IND_GET_CTX_PARAM(((KDF_SSHKDF *)vctx), params))
        ret = 1;
    else
        return 0;
    return ret;
}
```

Reply:
The pattern used within these get/set functions is that if a parameter is unknown then it just ignores the parameter. This is so new parameters dont cause an error in older providers.

So if the param[] has one unknown field followed by a terminator then it returns 1. If params is NULL we also just return 1.  So a param[] with just a terminator should just return 1 also.

Reply:
So if I add some indicators for KDF, I should modify this code snippet in the [test/evp_kdf_test.c](https://github.com/openssl/openssl/blob/8f250985ad1ac4efc25621ce2504c52ef0cbe283/test/evp_kdf_test.c#L276-L280) file, right?

```diff
diff --git a/test/evp_kdf_test.c b/test/evp_kdf_test.c
index 67de1cc7a7..d45d7277bb 100644
--- a/test/evp_kdf_test.c
+++ b/test/evp_kdf_test.c
@@ -273,9 +273,9 @@ static int do_kdf_hkdf_gettables(int expand_only, int has_digest)
             goto err;
     }
 
-    /* Get params returns -2 if an unsupported parameter is requested */
+    /* Get params returns 1 if an unsupported parameter is requested */
     params_get[0] = OSSL_PARAM_construct_end();
-    if (!TEST_int_eq(EVP_KDF_CTX_get_params(kctx, params_get), -2))
+    if (!TEST_int_eq(EVP_KDF_CTX_get_params(kctx, params_get), 1))
         goto err;
     ret = 1;
 err
```

Reply:
I would prefer not to continue to use this PR to have a discussion about another PR.

Reply:
Okay, I will start new discussion in a dedicated PR. Thank you.